File: frontend/services/admin/avaliacoes.ts
import { apiAuth } from "../api";
import type { Questao } from "./questoes";

export const STATUS_LABELS: Record<string, string> = {
  rascunho:  "Rascunho",
  publicada: "Publicada",
  encerrada: "Encerrada",
};

export const STATUS_COLORS: Record<string, string> = {
  rascunho:  "bg-border text-muted",
  publicada: "bg-emerald-100 text-emerald-700",
  encerrada: "bg-rose-100 text-rose-600",
};

export const STATUS_ALUNO_LABELS: Record<string, string> = {
  aguardando_correcao: "Aguardando correção",
  concluida:           "Concluída",
};

export interface AvaliacaoQuestaoItem {
  id: number;
  questao_id: number;
  ordem: number;
  peso: number;
  questao: Questao;
}

export interface AvaliacaoList {
  id: number;
  titulo: string;
  topicos: string[];
  modulo: string | null;
  tipo: string;
  turma_id: number;
  turma_nome: string | null;
  aula_id: number | null;
  data_aplicacao: string | null;
  hora_inicio: string | null;
  hora_fim: string | null;
  status: string;
  criado_em: string;
  total_questoes: number;
  total_alunos: number;
}

export interface AvaliacaoDetail extends AvaliacaoList {
  descricao: string | null;
  criado_por_id: number | null;
  total_pendentes: number;
  questoes: AvaliacaoQuestaoItem[];
  aula_data: string | null;
  aula_hora_inicio: string | null;
  aula_hora_fim: string | null;
}

export interface AvaliacaoCreate {
  titulo: string;
  topicos: string[];
  modulo?: string;
  descricao?: string;
  tipo?: string;
  turma_id: number;
  aula_id?: number;
  data_aplicacao?: string;
  hora_inicio?: string;
  hora_fim?: string;
  questoes?: { questao_id: number; ordem: number; peso: number }[];
}

export interface AvaliacaoAluno {
  aluno_id: number;
  nome: string;
  status: string;
  nota_final: number | null;
  iniciado_em: string;
  concluido_em: string | null;
}

export interface AvaliacaoDesempenhoAluno {
  id: number;
  titulo: string;
  turma_id: number;
  turma_nome: string | null;
  data_aplicacao: string | null;
  status: string;
  total_questoes: number;
  status_aluno: string;
  nota_final: number | null;
}

export interface RespostaAluno {
  questao_id: number;
  codigo: string;
  enunciado: string;
  subtipo: string;
  peso: number;
  resposta_dada: string | null;
  acertou: boolean | null;
  nota_manual: number | null;
  comentario_professor: string | null;
  corrigida: boolean;
}

export const avaliacoesService = {
  listar: (params?: {
    turma_id?: number;
    professor_id?: number;
    status?: string;
    topico?: string;
    modulo?: string;
    data_inicio?: string;
    data_fim?: string;
    page?: number;
    page_size?: number;
  }) => {
    const qs = new URLSearchParams();
    if (params?.turma_id) qs.set("turma_id", String(params.turma_id));
    if (params?.professor_id) qs.set("professor_id", String(params.professor_id));
    if (params?.status) qs.set("status", params.status);
    if (params?.topico) qs.set("topico", params.topico);
    if (params?.modulo) qs.set("modulo", params.modulo);
    if (params?.data_inicio) qs.set("data_inicio", params.data_inicio);
    if (params?.data_fim) qs.set("data_fim", params.data_fim);
    if (params?.page) qs.set("page", String(params.page));
    if (params?.page_size) qs.set("page_size", String(params.page_size));
    return apiAuth.get<AvaliacaoList[]>(`/avaliacoes/${qs.toString() ? `?${qs}` : ""}`);
  },

  buscar: (id: number) => apiAuth.get<AvaliacaoDetail>(`/avaliacoes/${id}`),

  criar: (data: AvaliacaoCreate) => apiAuth.post<AvaliacaoDetail>("/avaliacoes/", data),

  atualizar: (id: number, data: Partial<AvaliacaoCreate>) =>
    apiAuth.patch<AvaliacaoDetail>(`/avaliacoes/${id}`, data),

  publicar: (id: number) => apiAuth.post<AvaliacaoDetail>(`/avaliacoes/${id}/publicar`, {}),

  encerrar: (id: number) => apiAuth.post<AvaliacaoDetail>(`/avaliacoes/${id}/encerrar`, {}),

  adicionarQuestao: (id: number, item: { questao_id: number; ordem: number; peso: number }) =>
    apiAuth.post<AvaliacaoDetail>(`/avaliacoes/${id}/questoes`, item),

  removerQuestao: (id: number, questaoId: number) =>
    apiAuth.delete<AvaliacaoDetail>(`/avaliacoes/${id}/questoes/${questaoId}`),

  lancarNotas: (id: number, notas: { aluno_id: number; nota: number | null }[]) =>
    apiAuth.post<void>(`/avaliacoes/${id}/lancar-notas`, { notas }),

  listarAlunos: (id: number) =>
    apiAuth.get<AvaliacaoAluno[]>(`/avaliacoes/${id}/alunos`),

  listarPorAluno: (alunoId: number) =>
    apiAuth.get<AvaliacaoDesempenhoAluno[]>(`/avaliacoes/por-aluno/${alunoId}`),

  respostasAluno: (id: number, alunoId: number) =>
    apiAuth.get<RespostaAluno[]>(`/avaliacoes/${id}/respostas/${alunoId}`),

  deletar: (id: number) => apiAuth.delete<void>(`/avaliacoes/${id}`),

  corrigir: (id: number, alunoId: number, questaoId: number, nota_manual: number, comentario?: string) =>
    apiAuth.patch<RespostaAluno>(`/avaliacoes/${id}/corrigir/${alunoId}/${questaoId}`, {
      nota_manual,
      comentario_professor: comentario,
    }),
};

File: backend/app/routers/avaliacoes.py
from typing import Optional

from fastapi import APIRouter, Depends, HTTPException, Query, status
from sqlalchemy.orm import Session

from app.database import get_db
from app.dependencies import require_staff_or_professor
from app.models.avaliacao import STATUS_AVALIACAO
from app.models.user import User
from app.repositories import avaliacao as repo
from app.schemas.avaliacao import (
    AvaliacaoAlunoOut,
    AvaliacaoCreate,
    AvaliacaoListOut,
    AvaliacaoOut,
    AvaliacaoQuestaoIn,
    AvaliacaoUpdate,
    CorrecaoIn,
    RespostaAlunoOut,
)

router = APIRouter(prefix="/avaliacoes", tags=["avaliacoes"])


def _get_or_404(db: Session, avaliacao_id: int):
    av = repo.buscar(db, avaliacao_id)
    if not av:
        raise HTTPException(status_code=404, detail="Avaliação não encontrada")
    return av


def _enrich(av, db: Session) -> dict:
    from app.models.avaliacao import AvaliacaoAluno
    total_alunos = db.query(AvaliacaoAluno).filter(AvaliacaoAluno.avaliacao_id == av.id).count()
    total_pendentes = db.query(AvaliacaoAluno).filter(
        AvaliacaoAluno.avaliacao_id == av.id,
        AvaliacaoAluno.status == "aguardando_correcao",
    ).count()
    turma_nome = av.turma.nome if av.turma else None
    return {
        **{c.name: getattr(av, c.name) for c in av.__table__.columns},
        "turma_nome": turma_nome,
        "questoes": av.questoes,
        "total_alunos": total_alunos,
        "total_pendentes": total_pendentes,
    }


@router.get("/", response_model=list[AvaliacaoListOut])
def listar(
    turma_id: Optional[int] = Query(None),
    status: Optional[str] = Query(None),
    topico: Optional[str] = Query(None),
    modulo: Optional[str] = Query(None),
    page: int = Query(1, ge=1),
    page_size: int = Query(20, ge=1, le=100),
    db: Session = Depends(get_db),
    _: User = Depends(require_staff_or_professor),
):
    from app.models.avaliacao import AvaliacaoAluno
    items, _ = repo.listar(
        db, turma_id=turma_id, status=status, topico=topico, modulo=modulo,
        skip=(page - 1) * page_size, limit=page_size,
    )
    result = []
    for av in items:
        total_alunos = db.query(AvaliacaoAluno).filter(AvaliacaoAluno.avaliacao_id == av.id).count()
        result.append(AvaliacaoListOut(
            id=av.id,
            titulo=av.titulo,
            topicos=av.topicos,
            modulo=av.modulo,
            turma_id=av.turma_id,
            turma_nome=av.turma.nome if av.turma else None,
            data_aplicacao=av.data_aplicacao,
            status=av.status,
            criado_em=av.criado_em,
            total_questoes=len(av.questoes),
            total_alunos=total_alunos,
        ))
    return result


@router.post("/", response_model=AvaliacaoOut, status_code=status.HTTP_201_CREATED)
def criar(
    dados: AvaliacaoCreate,
    db: Session = Depends(get_db),
    current_user: User = Depends(require_staff_or_professor),
):
    av = repo.criar(db, dados, criado_por_id=current_user.id)
    return AvaliacaoOut(**_enrich(av, db))


@router.get("/{avaliacao_id}", response_model=AvaliacaoOut)
def buscar(
    avaliacao_id: int,
    db: Session = Depends(get_db),
    _: User = Depends(require_staff_or_professor),
):
    av = _get_or_404(db, avaliacao_id)
    return AvaliacaoOut(**_enrich(av, db))


@router.patch("/{avaliacao_id}", response_model=AvaliacaoOut)
def atualizar(
    avaliacao_id: int,
    dados: AvaliacaoUpdate,
    db: Session = Depends(get_db),
    _: User = Depends(require_staff_or_professor),
):
    av = _get_or_404(db, avaliacao_id)
    if av.status != "rascunho":
        raise HTTPException(status_code=409, detail="Só é possível editar avaliações em rascunho")
    av = repo.atualizar(db, av, dados)
    return AvaliacaoOut(**_enrich(av, db))


@router.post("/{avaliacao_id}/publicar", response_model=AvaliacaoOut)
def publicar(
    avaliacao_id: int,
    db: Session = Depends(get_db),
    _: User = Depends(require_staff_or_professor),
):
    av = _get_or_404(db, avaliacao_id)
    if av.status != "rascunho":
        raise HTTPException(status_code=409, detail="Avaliação já foi publicada ou encerrada")
    if not av.questoes:
        raise HTTPException(status_code=422, detail="Adicione ao menos uma questão antes de publicar")
    av = repo.mudar_status(db, av, "publicada")
    return AvaliacaoOut(**_enrich(av, db))


@router.post("/{avaliacao_id}/encerrar", response_model=AvaliacaoOut)
def encerrar(
    avaliacao_id: int,
    db: Session = Depends(get_db),
    _: User = Depends(require_staff_or_professor),
):
    av = _get_or_404(db, avaliacao_id)
    if av.status != "publicada":
        raise HTTPException(status_code=409, detail="Só é possível encerrar avaliações publicadas")
    av = repo.mudar_status(db, av, "encerrada")
    return AvaliacaoOut(**_enrich(av, db))


@router.post("/{avaliacao_id}/questoes", response_model=AvaliacaoOut)
def adicionar_questao(
    avaliacao_id: int,
    item: AvaliacaoQuestaoIn,
    db: Session = Depends(get_db),
    _: User = Depends(require_staff_or_professor),
):
    av = _get_or_404(db, avaliacao_id)
    if av.status != "rascunho":
        raise HTTPException(status_code=409, detail="Só é possível editar questões em rascunho")
    repo.adicionar_questao(db, avaliacao_id, item)
    db.refresh(av)
    return AvaliacaoOut(**_enrich(av, db))


@router.delete("/{avaliacao_id}/questoes/{questao_id}", response_model=AvaliacaoOut)
def remover_questao(
    avaliacao_id: int,
    questao_id: int,
    db: Session = Depends(get_db),
    _: User = Depends(require_staff_or_professor),
):
    av = _get_or_404(db, avaliacao_id)
    if av.status != "rascunho":
        raise HTTPException(status_code=409, detail="Só é possível editar questões em rascunho")
    repo.remover_questao(db, avaliacao_id, questao_id)
    db.refresh(av)
    return AvaliacaoOut(**_enrich(av, db))


@router.get("/{avaliacao_id}/alunos", response_model=list[AvaliacaoAlunoOut])
def listar_alunos(
    avaliacao_id: int,
    db: Session = Depends(get_db),
    _: User = Depends(require_staff_or_professor),
):
    from app.models.avaliacao import AvaliacaoAluno
    from app.models.pessoa import Pessoa
    registros = (
        db.query(AvaliacaoAluno, Pessoa)
        .join(Pessoa, Pessoa.id == AvaliacaoAluno.aluno_id)
        .filter(AvaliacaoAluno.avaliacao_id == avaliacao_id)
        .order_by(AvaliacaoAluno.iniciado_em.desc())
        .all()
    )
    return [
        AvaliacaoAlunoOut(
            aluno_id=reg.aluno_id,
            nome=pessoa.nome,
            status=reg.status,
            nota_final=reg.nota_final,
            iniciado_em=reg.iniciado_em,
            concluido_em=reg.concluido_em,
        )
        for reg, pessoa in registros
    ]


@router.get("/{avaliacao_id}/respostas/{aluno_id}", response_model=list[RespostaAlunoOut])
def respostas_aluno(
    avaliacao_id: int,
    aluno_id: int,
    db: Session = Depends(get_db),
    _: User = Depends(require_staff_or_professor),
):
    av = _get_or_404(db, avaliacao_id)
    peso_map = {aq.questao_id: aq.peso for aq in av.questoes}
    respostas = repo.buscar_respostas_aluno(db, avaliacao_id, aluno_id)
    result = []
    for r in respostas:
        result.append(RespostaAlunoOut(
            questao_id=r.questao_id,
            codigo=r.questao.codigo,
            enunciado=r.questao.enunciado,
            subtipo=r.questao.subtipo,
            peso=peso_map.get(r.questao_id, 1.0),
            resposta_dada=r.resposta_dada,
            acertou=r.acertou,
            nota_manual=r.nota_manual,
            comentario_professor=r.comentario_professor,
            corrigida=r.corrigida,
        ))
    return result


@router.patch("/{avaliacao_id}/corrigir/{aluno_id}/{questao_id}", response_model=RespostaAlunoOut)
def corrigir(
    avaliacao_id: int,
    aluno_id: int,
    questao_id: int,
    dados: CorrecaoIn,
    db: Session = Depends(get_db),
    _: User = Depends(require_staff_or_professor),
):
    av = _get_or_404(db, avaliacao_id)
    peso_map = {aq.questao_id: aq.peso for aq in av.questoes}
    try:
        resp = repo.corrigir_resposta(db, avaliacao_id, aluno_id, questao_id, dados.nota_manual, dados.comentario_professor)
    except ValueError as e:
        raise HTTPException(status_code=404, detail=str(e))
    return RespostaAlunoOut(
        questao_id=resp.questao_id,
        codigo=resp.questao.codigo,
        enunciado=resp.questao.enunciado,
        subtipo=resp.questao.subtipo,
        peso=peso_map.get(resp.questao_id, 1.0),
        resposta_dada=resp.resposta_dada,
        acertou=resp.acertou,
        nota_manual=resp.nota_manual,
        comentario_professor=resp.comentario_professor,
        corrigida=resp.corrigida,
    )
